### DISCUSSION TITLE
Help converting an HTML span to a Markdown directive

### DISCUSSION THREAD

**Discussion by dmca-glasgow** [unifiedjs/unified#240] [Q&A] [answered]
Hi there,

In the past I have used unified to convert Markdown to HTML, and when I made a custom transformation, I used the hast properties (hName, hProperties, hChildren) in the mdast AST.

However, I'm now tasked with converting HTML to Markdown and there doesn't seem to be a similar way to do this?

My specifc problem is converting a span with class to a [leaf directive](https://github.com/remarkjs/remark-directive):

```html
<p>My content and <span class="sidenote">my <b>sidenote</b></span>.</p>
```

should become:

```md
My content and ::sidenote[my **sidenote**].
```

Where in this processor should I put my custom transformation, and what would it look like?

```ts
import { unified } from 'unified';
import rehypeParse from 'rehype-parse';
import rehypeRemark from 'rehype-remark';
import remarkStringify from 'remark-stringify';

const file = await unified()
  .use(rehypeParse)
  .use(rehypeRemark)
  .use(remarkStringify)
  .process(`<p>My content and <span class="sidenote">my <b>sidenote</b></span>.</p>`)

console.log(String(file))

// current output is:
// My content and my **sidenote**.
```

Thanks for any help.

**Comment by wooorm** [comment, marked as answer]
Hey!

The h* fields are indeed also a good way to do this, the other way around, mdast -> hast.
Such a thing does not exist hast -> mdast.

But, what is also possible, is to hook into the place where the conversion between the ASTs happens. `remark-rehype` and `rehype-remark`.

They both support [handlers](https://github.com/rehypejs/rehype-remark#options):

> `handlers` (`Record<string, Handle>`, optional) — object mapping tag names to functions handling the corresponding elements; merged into the defaults; see [Handle in `hast-util-to-mdast`](https://github.com/syntax-tree/hast-util-to-mdast#handle)

There you would use `span` as a key, and write a handler that generates a text directive: https://github.com/syntax-tree/mdast-util-directive#textdirective.

**Reply by dmca-glasgow** [reply]
Thanks for your help @wooorm,

I ended up with something like this:

```ts
import { unified } from 'unified';
import rehypeParse from 'rehype-parse';
import rehypeRemark from 'rehype-remark';
import remarkStringify from 'remark-stringify';
import remarkDirective from 'remark-directive';
import { toMdast, State } from 'hast-util-to-mdast';
import { TextDirective } from 'mdast-util-directive';
import { Parent, PhrasingContent } from 'mdast';
import { Element } from 'hast';

const file = await unified()
  .use(rehypeParse)
  .use(rehypeRemark, {
    handlers: {
      span: customSpanHandler,
    },
  })
  .use(remarkDirective)
  .use(remarkStringify)
  .process(`<p>My content and <span class="sidenote">my <b>sidenote</b></span>.</p>`);

function customSpanHandler(state: State, node: Element) {
  if (node.properties.className?.toString() === 'sidenote') {

    const root = toMdast({
      type: 'root',
      children: node.children,
    }) as Parent;

    const result: TextDirective = {
      type: 'textDirective',
      name: 'sidenote',
      children: root.children as PhrasingContent[],
    };

    state.patch(node, result);
    return result;
  }

  return state.all(node);
}

console.log(String(file));
// My content and :sidenote[my **sidenote**].
```

What stumped me for quite a while was that `leafDirective` has to have it's own line, and `textDirective` is more suitable for the 'inline' scenario (with just the one `:`).

If anything can be improved please let me know as I have to write a lot of these!

**Reply by wooorm** [reply]
Looks good, main thing is I wouldn’t call `toMdast` in there. You can use `state` to do that. See:
https://github.com/syntax-tree/hast-util-to-mdast/blob/52b3d5a715da233d15b711c5475d1cf40480a7cc/lib/handlers/a.js for example.

And, I’d go with a different condition. So, something along the lines of this pseudo code:

```ts
function customSpanHandler(state: State, node: Element) {
  if (
    Array.isArray(node.properties.className) &&
    node.properties.className.includes('sidenote')
  ) {
    const result: TextDirective = {
      type: 'textDirective',
      name: 'sidenote',
      children: state.all(node) as Array<PhrasingContent>
    }

    state.patch(node, result)
    return result
  }

  return state.all(node)
}
```

**Reply by dmca-glasgow** [reply]
Great, thanks for the useful suggestions!
